rust/parquet/src/arrow/schema.rs

```diff
@@ -591,6 +591,7 @@ impl ParquetTypeConverter<'_> {
             LogicalType::INT_32 => Ok(DataType::Int32),
             LogicalType::DATE => Ok(DataType::Date32(DateUnit::Day)),
             LogicalType::TIME_MILLIS => Ok(DataType::Time32(TimeUnit::Millisecond)),
+            LogicalType::DECIMAL => self.to_decimal(),
```

Comment:
should this be handled somehow in the parquet/arrow reader?

Reply:
I am having a hard time finding the right place to construct the decimal type. My main concern with the changes I made is that Decimal is the only type that needs additional info from the schema in order to define precision and scale - all other types can be built directly from the logical type. However, if I put it in `PrimitiveArrayReader::new` I am only able to accomplish this by reimplementing parts of the ParquetTypeConverter specifically for Decimal. Do you have a suggestion how I can handle this in a better way?

Reply:
Hmm, should this be handled in `build_for_primitive_type_inner` inside `array_reader.rs`? where we construct a `ComplexObjectArrayReader` with the right converter for decimal types. I see we currently handle the case for `FIXED_LEN_BYTE_ARRAY` but not for `INT32`/`INT64`.

Reply:
I added an implementation using converters, but the creation of `ParquetTypeConverter` occurs before we even get to `build_for_primitive_type_inner`:

```
ParquetFileArrowReader::get_record_reader ->
ParquetFileArrowReader::get_record_reader_by_columns -> 
ParquetFileArrowReader::get_schema ->
schema::parquet_to_arrow_schema ->
schema::parquet_to_arrow_schema_by_columns ->
ParquetTypeConverter::new
```
Thus, if we want to avoid the decimal case, a decimal specific implementation in `get_schema` would probably be required.

The converters could avoid the decimal specific branch in PrimitiveArrayReader::next_batch:

```
ArrowType::Decimal(p, s) => {
                let to_int64 = arrow::compute::cast(&array, &ArrowType::Int64)?;
                let mut builder = DecimalBuilder::new(to_int64.len(), *p, *s);
                let values = to_int64.as_any().downcast_ref::<Int64Array>().unwrap();
                for maybe_value in values.iter() {
                    match maybe_value {
                        Some(value) => builder.append_value(value as i128)?,
                        None => builder.append_null()?
                    }
                }
                Arc::new(builder.finish()) as ArrayRef
            }
```

We could avoid the cast to i64 by adding converters for i32/i64 for but I think this is not such a big deal - what do you think?

Reply:
Got it. This looks good but why we need to do another cast on the arrow array though? it could hurt performance. Is it possible to directly build decimal array from the input parquet data?

Reply:
@sweb what does downcasting the array straignt to an i64 result in? Does the below panic?

```rust
let mut builder = DecimalBuilder::new(array.len(), *p, *s);
let values = array.as_any().downcast_ref::<Int64Array>().unwrap();
```

Reply:
The cast is necessary, because downcasting an i32 array to i64 results in None, panicking as a result. I have removed the cast and handle the building of the decimal array now separately, but I do not really like my solution:

```
                let mut builder = DecimalBuilder::new(array.len(), *p, *s);
                match array.data_type() {
                    ArrowType::Int32 => {
                        let values = array.as_any().downcast_ref::<Int32Array>().unwrap();
                        for maybe_value in values.iter() {
                            match maybe_value {
                                Some(value) => builder.append_value(value as i128)?,
                                None => builder.append_null()?,
                            }
                        }
                    }
                    ArrowType::Int64 => {
                        let values = array.as_any().downcast_ref::<Int64Array>().unwrap();
                        for maybe_value in values.iter() {
                            match maybe_value {
                                Some(value) => builder.append_value(value as i128)?,
                                None => builder.append_null()?,
                            }
                        }
                    }
                    _ => {
                        return Err(ArrowError(format!(
                            "Cannot convert {:?} to decimal",
                            array.data_type()
                        )))
                    }
                }
```

At the moment, I am not able to convince the compiler to treat downcasts to i32 and i64 as the same primitive array and then use `as` on the resulting iterator to convert from types that are unknown at compile time to `i128`. I will probably need to spend some time figuring this out.

Alternatively, we could use converters, similar to the fixed length case.

Reply:
Thanks for the update. This PR looks good and we can address the remaining in a followup.

Reply:
@sunchao thank you for your help with this PR. The result is much better than my initial proposal!

Reply:
np @sweb - thanks for the contribution!